File: src/routes/pengurus.ts
import { Hono } from "hono";
import { Env, Divisi, Pengurus } from "../types";
import { successResponse, errorResponse } from "../utils/response";
import { authMiddleware } from "../middleware/auth";

const pengurusRouter = new Hono<{ Bindings: Env }>();

/**
 * GET /api/pengurus (Public)
 * Returns all divisions with their nested board members
 */
pengurusRouter.get("/", async (c) => {
    const { results: divisions } = await c.env.DB.prepare(
        "SELECT * FROM divisi ORDER BY order_priority ASC"
    ).all<Divisi>();

    const { results: members } = await c.env.DB.prepare(
        "SELECT * FROM pengurus ORDER BY order_priority ASC"
    ).all<Pengurus>();

    const structured = (divisions || []).map(div => ({
        ...div,
        members: (members || []).filter(m => m.divisi_id === div.id)
    }));

    return c.json(successResponse(structured));
});

/**
 * POST /api/pengurus/divisi (Protected - Admin)
 */
pengurusRouter.post("/divisi", authMiddleware, async (c) => {
    const { name, description, order_priority } = await c.req.json<{
        name?: string;
        description?: string;
        order_priority?: number;
    }>();

    if (!name) {
        return c.json(errorResponse("Nama divisi wajib diisi"), 400);
    }

    const id = `div_${Date.now()}_${Math.random().toString(36).substring(2, 6)}`;
    await c.env.DB.prepare(
        "INSERT INTO divisi (id, name, description, order_priority) VALUES (?, ?, ?, ?)"
    ).bind(id, name, description || null, order_priority || 0).run();

    const created = await c.env.DB.prepare("SELECT * FROM divisi WHERE id = ?").bind(id).first<Divisi>();
    return c.json(successResponse(created, "Divisi berhasil dibuat"), 201);
});

/**
 * POST /api/pengurus/member (Protected - Admin)
 */
pengurusRouter.post("/member", authMiddleware, async (c) => {
    const { divisi_id, name, role, photo_url, period, order_priority } = await c.req.json<{
        divisi_id?: string;
        name?: string;
        role?: string;
        photo_url?: string;
        period?: string;
        order_priority?: number;
    }>();

    if (!divisi_id || !name || !role || !period) {
        return c.json(errorResponse("Field divisi_id, name, role, dan period wajib diisi"), 400);
    }

    const id = `png_${Date.now()}_${Math.random().toString(36).substring(2, 6)}`;
    await c.env.DB.prepare(
        "INSERT INTO pengurus (id, divisi_id, name, role, photo_url, period, order_priority) VALUES (?, ?, ?, ?, ?, ?, ?)"
    ).bind(id, divisi_id, name, role, photo_url || null, period, order_priority || 0).run();

    const created = await c.env.DB.prepare("SELECT * FROM pengurus WHERE id = ?").bind(id).first<Pengurus>();
    return c.json(successResponse(created, "Pengurus berhasil ditambahkan"), 201);
});

/**
 * DELETE /api/pengurus/member/:id (Protected - Admin)
 */
pengurusRouter.delete("/member/:id", authMiddleware, async (c) => {
    const id = c.req.param("id");
    await c.env.DB.prepare("DELETE FROM pengurus WHERE id = ?").bind(id).run();
    return c.json(successResponse(null, "Pengurus berhasil dihapus"));
});

export default pengurusRouter;

File: routers/pengurus.py
import time
import random
import string
from typing import List
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from database import get_db
import models
import schemas
from utils.auth import get_current_admin

router = APIRouter(prefix="/api/pengurus", tags=["Divisi & Pengurus"])

@router.get("", response_model=schemas.ApiResponse[List[schemas.DivisiOut]])
def get_all_divisions_and_members(db: Session = Depends(get_db)):
    """Get all divisions with nested members."""
    divisions = db.query(models.Divisi).order_by(models.Divisi.order_priority.asc()).all()
    return schemas.ApiResponse(
        success=True,
        data=[schemas.DivisiOut.model_validate(div) for div in divisions]
    )

@router.post("/divisi", response_model=schemas.ApiResponse[schemas.DivisiOut], status_code=201)
def create_divisi(
    payload: schemas.DivisiCreate,
    current_user: models.User = Depends(get_current_admin),
    db: Session = Depends(get_db)
):
    """Create new division (Admin only)."""
    div_id = payload.id
    if not div_id:
        random_suffix = ''.join(random.choices(string.ascii_lowercase + string.digits, k=4))
        div_id = f"div_{int(time.time())}_{random_suffix}"

    divisi = models.Divisi(
        id=div_id,
        name=payload.name,
        komisi=payload.komisi,
        icon_name=payload.icon_name,
        description=payload.description,
        group_photo_url=payload.group_photo_url,
        order_priority=payload.order_priority or 0
    )
    db.add(divisi)
    db.commit()
    db.refresh(divisi)

    return schemas.ApiResponse(
        success=True,
        data=schemas.DivisiOut.model_validate(divisi),
        message="Divisi berhasil dibuat"
    )

@router.put("/divisi/{divisi_id}", response_model=schemas.ApiResponse[schemas.DivisiOut])
def update_divisi(
    divisi_id: str,
    payload: schemas.DivisiUpdate,
    current_user: models.User = Depends(get_current_admin),
    db: Session = Depends(get_db)
):
    """Update division (Admin only)."""
    divisi = db.query(models.Divisi).filter(models.Divisi.id == divisi_id).first()
    if not divisi:
        raise HTTPException(status_code=404, detail="Divisi tidak ditemukan")

    if payload.name is not None:
        divisi.name = payload.name
    if payload.komisi is not None:
        divisi.komisi = payload.komisi
    if payload.icon_name is not None:
        divisi.icon_name = payload.icon_name
    if payload.description is not None:
        divisi.description = payload.description
    if payload.group_photo_url is not None:
        divisi.group_photo_url = payload.group_photo_url
    if payload.order_priority is not None:
        divisi.order_priority = payload.order_priority

    db.commit()
    db.refresh(divisi)

    return schemas.ApiResponse(
        success=True,
        data=schemas.DivisiOut.model_validate(divisi),
        message="Divisi berhasil diperbarui"
    )

@router.delete("/divisi/{divisi_id}", response_model=schemas.ApiResponse[None])
def delete_divisi(
    divisi_id: str,
    current_user: models.User = Depends(get_current_admin),
    db: Session = Depends(get_db)
):
    """Delete division (Admin only)."""
    divisi = db.query(models.Divisi).filter(models.Divisi.id == divisi_id).first()
    if not divisi:
        raise HTTPException(status_code=404, detail="Divisi tidak ditemukan")

    db.delete(divisi)
    db.commit()
    return schemas.ApiResponse(success=True, message="Divisi berhasil dihapus")

@router.post("/member", response_model=schemas.ApiResponse[schemas.PengurusOut], status_code=201)
def create_member(
    payload: schemas.PengurusCreate,
    current_user: models.User = Depends(get_current_admin),
    db: Session = Depends(get_db)
):
    """Create new board member (Admin only)."""
    divisi = db.query(models.Divisi).filter(models.Divisi.id == payload.divisi_id).first()
    if not divisi:
        raise HTTPException(status_code=404, detail="Divisi tidak ditemukan")

    random_suffix = ''.join(random.choices(string.ascii_lowercase + string.digits, k=4))
    member_id = f"png_{int(time.time())}_{random_suffix}"

    member = models.Pengurus(
        id=member_id,
        divisi_id=payload.divisi_id,
        name=payload.name,
        role=payload.role,
        photo_url=payload.photo_url,
        period=payload.period or "2025/2026",
        order_priority=payload.order_priority or 0
    )
    db.add(member)
    db.commit()
    db.refresh(member)

    return schemas.ApiResponse(
        success=True,
        data=schemas.PengurusOut.model_validate(member),
        message="Pengurus berhasil ditambahkan"
    )

@router.put("/member/{member_id}", response_model=schemas.ApiResponse[schemas.PengurusOut])
def update_member(
    member_id: str,
    payload: schemas.PengurusUpdate,
    current_user: models.User = Depends(get_current_admin),
    db: Session = Depends(get_db)
):
    """Update board member (Admin only)."""
    member = db.query(models.Pengurus).filter(models.Pengurus.id == member_id).first()
    if not member:
        raise HTTPException(status_code=404, detail="Pengurus tidak ditemukan")

    if payload.divisi_id is not None:
        member.divisi_id = payload.divisi_id
    if payload.name is not None:
        member.name = payload.name
    if payload.role is not None:
        member.role = payload.role
    if payload.photo_url is not None:
        member.photo_url = payload.photo_url
    if payload.period is not None:
        member.period = payload.period
    if payload.order_priority is not None:
        member.order_priority = payload.order_priority

    db.commit()
    db.refresh(member)

    return schemas.ApiResponse(
        success=True,
        data=schemas.PengurusOut.model_validate(member),
        message="Data pengurus berhasil diperbarui"
    )

@router.delete("/member/{member_id}", response_model=schemas.ApiResponse[None])
def delete_member(
    member_id: str,
    current_user: models.User = Depends(get_current_admin),
    db: Session = Depends(get_db)
):
    """Delete board member (Admin only)."""
    member = db.query(models.Pengurus).filter(models.Pengurus.id == member_id).first()
    if not member:
        raise HTTPException(status_code=404, detail="Pengurus tidak ditemukan")

    db.delete(member)
    db.commit()
    return schemas.ApiResponse(success=True, message="Pengurus berhasil dihapus")
